### PR TITLE
Share memory in PutRequest in server

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
We've done the same thing frontend with GetResponse, this is the counterpart for PutRequest in server.

### CODE REVIEW

File: ambry-utils/src/test/java/com.github.ambry.utils/UtilsTest.java [hunk 1]
```diff
@@ -192,6 +194,75 @@ public void testReadBuffers() throws IOException {
...
+    buffer.flip();
+    byte[] expectedArray = firstRandomBytes;
+    while (expectedArray != null) {
+      CrcInputStream cis = new CrcInputStream(new ByteBufferDataInputStream(buffer));
+      DataInputStream dis = new DataInputStream(cis);
+      long dataSize = dis.readLong();
+      assertEquals((long) dataSize, blobSize);
+      ByteBufferInputStream obtained = Utils.getByteBufferInputStreamFromCRCInputStream(cis, (int) dataSize);
+      // Make sure these two ByteBuffers actually share the underlying memory.
+      assertEquals(getByteArrayFromByteBuffer(buffer), getByteArrayFromByteBuffer(obtained.getByteBuffer()));
```
Comment (lightningrob) [inline ambry-utils/src/test/java/com.github.ambry.utils/UtilsTest.java]: How does the below code ensure that the memory is shared, as opposed to the byte arrays being equals?

File: ambry-utils/src/main/java/com.github.ambry.utils/Utils.java [hunk 2]
```diff
@@ -210,6 +211,64 @@ public static ByteBuffer readShortBuffer(DataInputStream input) throws IOExcepti
     return buffer;
   }
 
+  /**
+   * Creating a {@link ByteBufferInputStream} from the {@link CrcInputStream} by sharing the underlying memory if the
```
Comment (lightningrob) [inline ambry-utils/src/main/java/com.github.ambry.utils/Utils.java]: Minor: "Create"

File: ambry-utils/src/main/java/com.github.ambry.utils/Utils.java [hunk 3]
```diff
@@ -210,6 +211,64 @@ public static ByteBuffer readShortBuffer(DataInputStream input) throws IOExcepti
...
+   * crcStream is built upon a {@link ByteBufferDataInputStream}.
+   * @param crcStream The crcStream to read {@link ByteBuffer} out.
+   * @param dataSize The size of {@link ByteBuffer}.
+   * @return The {@link ByteBufferInputStream}
+   * @throws IOException Unexpected IO errors.
+   */
+  public static ByteBufferInputStream getByteBufferInputStreamFromCRCInputStream(CrcInputStream crcStream, int dataSize)
```
Comment (lightningrob) [inline ambry-utils/src/main/java/com.github.ambry.utils/Utils.java]: CRC -> Crc

File: ambry-utils/src/main/java/com.github.ambry.utils/Utils.java [hunk 4]
```diff
@@ -210,6 +211,64 @@ public static ByteBuffer readShortBuffer(DataInputStream input) throws IOExcepti
...
+      byteBuffer.limit(oldLimit);
+      // Change the byte buffer's position as if the data is fetched.
+      byteBuffer.position(startIndex + dataSize);
+    } else {
+      output = new ByteBufferInputStream(crcStream, dataSize);
+    }
+    return output;
+  }
+
+  public static ByteBuffer readByteBufferFromCRCInputStream(CrcInputStream crcStream, int dataSize) throws IOException {
```
Comment (lightningrob) [inline ambry-utils/src/main/java/com.github.ambry.utils/Utils.java]: CRC -> Crc.  Also javadoc.

File: ambry-utils/src/main/java/com.github.ambry.utils/Utils.java [hunk 5]
```diff
@@ -210,6 +211,64 @@ public static ByteBuffer readShortBuffer(DataInputStream input) throws IOExcepti
...
+    ByteBuffer result;
+    InputStream inputStream = crcStream.getUnderlyingInputStream();
```
Comment (lightningrob) [inline ambry-utils/src/main/java/com.github.ambry.utils/Utils.java]: Lot of duplication in next 9 lines compared to previous method.  Possible to reuse?

File: ambry-utils/src/test/java/com.github.ambry.utils/UtilsTest.java [hunk 6]
```diff
@@ -192,6 +194,75 @@ public void testReadBuffers() throws IOException {
     }
   }
 
+  @Test
+  public void testBlobDataShareMemory() throws Exception {
```
Comment (lightningrob) [inline ambry-utils/src/test/java/com.github.ambry.utils/UtilsTest.java]: Rename this.

File: ambry-utils/src/test/java/com.github.ambry.utils/UtilsTest.java [hunk 7]
```diff
@@ -192,6 +194,75 @@ public void testReadBuffers() throws IOException {
...
+      byte[] obtainedArray = new byte[blobSize];
+      obtained.read(obtainedArray);
+      assertArrayEquals(obtainedArray, expectedArray);
+      long crcRead = buffer.getLong();
+      assertEquals(crcRead, cis.getValue());
+      if (expectedArray == firstRandomBytes) {
+        expectedArray = secondRandomBytes;
+      } else {
+        expectedArray = null;
+      }
```
Comment (lightningrob) [inline ambry-utils/src/test/java/com.github.ambry.utils/UtilsTest.java]: expectedArray = (expectedArray == firstRandomBytes) ? secondRandomBytes : null;

File: ambry-utils/src/test/java/com.github.ambry.utils/UtilsTest.java (hunk 1 above)
Comment (justinlin-linkedin) [reply ambry-utils/src/test/java/com.github.ambry.utils/UtilsTest.java]: Byte array being equal means these two ByteBuffers use the same byte array underlying. Here sharing memory means we are not creating a new byte array with the same bytes in it.

File: ambry-utils/src/main/java/com.github.ambry.utils/Utils.java [hunk 8]
```diff
@@ -210,6 +211,79 @@ public static ByteBuffer readShortBuffer(DataInputStream input) throws IOExcepti
...
+
+  /**
+   * Transfer {@code dataSize} bytes of data from the given crc stream to a newly create {@link ByteBuffer}. The method
+   * would also update the crc value in the crc stream.
+   * @param crcStream The crc stream.
+   * @param dataSize The number of bytes to transfer.
+   * @return the newly created {@link ByteBuffer} which contains the transferred data.
+   * @throws IOException Any I/O error.
+   */
+  public static ByteBuffer readByteBufferFromCrcInputStream(CrcInputStream crcStream, int dataSize) throws IOException {
```
Comment (zzmao) [inline ambry-utils/src/main/java/com.github.ambry.utils/Utils.java]: Also add a test for this method?

Comment (justinlin-linkedin) [reply ambry-utils/src/main/java/com.github.ambry.utils/Utils.java]: I kinda merged this function with getByteBufferInputStreamFromCrcInputStream. I don't think we need to write another test case here.

**Pull request merged**